sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
 }
 
 // <id_scope>/registrations/<registration_id>/api-version=<service_version>
-AZ_NODISCARD az_result az_iot_provisioning_client_user_name_get(
+AZ_NODISCARD az_result az_iot_provisioning_client_get_user_name(
     az_iot_provisioning_client const* client,
-    az_span mqtt_user_name,
-    az_span* out_mqtt_user_name)
+    char* mqtt_user_name,
+    size_t mqtt_user_name_size,
+    size_t* out_mqtt_user_name_length)
 {
   AZ_PRECONDITION_NOT_NULL(client);
-  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+  AZ_PRECONDITION_NOT_NULL(mqtt_user_name);
+  AZ_PRECONDITION(mqtt_user_name_size > 0);
+
+  az_span mqtt_user_name_span
+      = az_span_init((uint8_t*)mqtt_user_name, (int32_t)mqtt_user_name_size);
 
   const az_span* const user_agent = &(client->_internal.options.user_agent);
+  az_span str_registrations
+      = az_span_slice(str_dps_registrations_res, idx_registrations_start, idx_registrations_end);
 
   int32_t required_length = az_span_size(client->_internal.id_scope)
       + az_span_size(str_registrations) + az_span_size(client->_internal.registration_id)
       + az_span_size(provisioning_service_api_version);
   if (az_span_size(*user_agent) > 0)
   {
-    required_length += az_span_size(*user_agent) + 1;
+    required_length += az_span_size(user_agent_version_prefix) + az_span_size(*user_agent);
```

Comment:
I wanted to ask this earlier, but is integer overflow possible here or is it very unlikely for anyone to be passing in such large buffers that `required_length` goes beyond int32 max given the domain is IoT devices? Is it OK for us to have undefined behavior in that case?

Reply:
It could in theory, very unlikely here since the values come from the application.
See https://github.com/Azure/azure-sdk-for-c/issues/340#issuecomment-588405246.